---
title: "IBS Colon samples analysis"
author: "Sushma Nagaraj"
date: "11/8/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
suppressPackageStartupMessages({
  library(here)
  library(DESeq2)
  library(PCAtools)
  library(stringr)
  library(reshape2)
  library(ggplot2)
  library(EDASeq)
  library(RUVSeq)
  library(RColorBrewer)
  library(IHW)
})
```

```{r}
dds.full <- readRDS(here("data", "dds.salmon_requantification_summarized_to_gene.rds"))
```

#Distal colon
```{r}
#Males and females combined
dds <- dds.full[ ,dds.full$source == "DISTAL COLON"]

#Females
dds <- dds.full[ ,dds.full$source == "DISTAL COLON" & dds.full$Gender2 == "F"]

#Males
dds <- dds.full[ ,dds.full$source == "DISTAL COLON" & dds.full$Gender2 == "M"]

table(dds$Gender2, dds$group)
```

```{r}
rowData(dds)$chr <- rowData(dds)$seqnames
rowData(dds) <- rowData(dds)[,-which(colnames(rowData(dds)) %in% c("seqnames", "ranges", "strand", "start", "end", "width"))] #These columns are throwing an error

rld <- rlog(dds)
p <- pca(assay(rld), metadata = colData(rld))
biplot(p, colby = "group", shape = "Gender2", lab = p$metadata$Sample, labSize = 3, pointSize = 3, legendPosition = 'right', title = "Control vs. IBS Rlog normalization")
biplot(p, colby = "Extraction.Date", shape = "group", lab = p$metadata$Sample, labSize = 3, pointSize = 3, legendPosition = 'right', title = "Extraction date")
biplot(p, colby = "Cage", shape = "group", lab = p$metadata$Sample, labSize = 3, pointSize = 3, legendPosition = 'right', title = "Cage")
biplot(p, colby = "Sac.Date", shape = "group", lab = p$metadata$Sample, labSize = 3, pointSize = 3, legendPosition = 'right', title = "Sac date")
```

#Filter pseudo genes, predicted genes, mitochondrial genes.
```{r , echo=FALSE}
pg <- grep("protein_coding", rowData(dds)$gene_type)
length(pg)
dds <- dds[pg, ]
mt <- which(rowData(dds)$chr == "chrM")
dds <- dds[-mt, ]
```


```{r}
set <- newSeqExpressionSet(as.matrix(counts(dds)), phenoData = data.frame(dds$group, row.names = colnames(counts(dds))), featureData = as.data.frame(rowData(dds)))

for(i in 1:ncol(set)){
  biasPlot(set[,c(1:i)], "gene_length", log=TRUE, ylim = c(0,10), main = colnames(set)[i])
}
```


```{r}
keep <- apply(counts(dds), 1, function(x) length(x[x>5]) >= 2)
dds <- dds[keep, ]
dds <- estimateSizeFactors(dds)
norm <- counts(dds, normalized = TRUE)
```


```{r}
colors <- brewer.pal(8, "Set2")
colLib <- colors[dds$group]
boxplot(log2(norm + 1), col = colLib, las = 2)
plotRLE(norm, col = colLib, outline = FALSE, ylim = c(-0.4, 0.4), las = 2)
plotPCA(norm, col = colLib, main = "DESeq2 normalization")

genes <- rownames(norm)
differences <- makeGroups(dds$group)
differences

ite <- min(8, nrow(colData(dds))-1) 
for(i in 1:ite) {
  n.k <- i
  s <- RUVs(round(norm), genes, k=n.k, differences)
  plotRLE(s$normalizedCounts, outline = FALSE, ylim = c(-2, 2), col=colLib, main = paste0("RUVs k=", n.k), las = 2)
  plotPCA(s$normalizedCounts, col=colLib, cex=1.2, main = paste0("RUVs k=", n.k))
}

n.k <- 8 #Distal colon all
n.k <- 2 #Distal colon males
n.k <- 5 #Distal colon females

s <- RUVs(round(norm), genes, k=n.k, differences)
plotRLE(s$normalizedCounts, outline = FALSE, ylim = c(-2, 2), col=colLib, main = paste0("RUVs k=", n.k), las = 2)
plotPCA(s$normalizedCounts, col=colLib, cex=1.2, main = paste0("RUVs k=", n.k))
s$W
all(rownames(s$W) == rownames(colData(dds)))
colData(dds) <- cbind(colData(dds), s$W)
d <- as.formula(paste0("~", paste(paste(paste0("W_", 1:n.k), collapse = " + "), "group", sep = " + ")))
d
design(dds) <- d

dds <- DESeq(dds)
res <- results(dds)
hist(res$pvalue)
summary(res)
DESeq2::plotMA(dds)

res$gene_name <- rowData(dds)$gene_name
resOrd <- res[order(res$pvalue), ]
ihwRes <- ihw(pvalue ~ baseMean,  data = as.data.frame(resOrd), alpha = 0.1)
resOrd$ihw <- adj_pvalues(ihwRes)
table(resOrd$ihw <= 0.1)
resOrd$geneID <- str_sub(rownames(resOrd), 1, 18)
result0.1 <- resOrd[resOrd$ihw <= 0.1, ]

plot(result0.1$log2FoldChange, -log10(result0.1$ihw), xlim = c(-2, 2), col = "red", pch = 16)
text(result0.1$log2FoldChange, -log10(result0.1$ihw), labels = result0.1$gene_name, cex = 0.8)

result15perfdr <- resOrd[resOrd$ihw <= 0.15, ] #For Control vs IBS
resup <- result15perfdr[result15perfdr$log2FoldChange > 0, ]
resdown <- result15perfdr[result15perfdr$log2FoldChange < 0, ]

group <- "Distal_colon"
gender <- "combined"
gender <- "Males"
gender <- "Females"

write.csv(result15perfdr, file = file.path("/data/users/sushmanagaraj/projects/rnaseq/IBS_GABA/analysis/Reanalysis_with_Salmon_quantification/IBS-vs-Control", group, gender, paste0(group, ".", gender, ".RUVs.DE_genes.padj15per.txt")))
write.csv(resOrd, file = file.path("/data/users/sushmanagaraj/projects/rnaseq/IBS_GABA/analysis/Reanalysis_with_Salmon_quantification/IBS-vs-Control", group, gender, paste0(group, ".", gender, ".RUVs.DE_analysis.txt")))
save(s, dds, file = file.path("/data/users/sushmanagaraj/projects/rnaseq/IBS_GABA/analysis/Reanalysis_with_Salmon_quantification/IBS-vs-Control", group, gender, "/norm_matrix_dds.RData"))
```


```{r}
colors <- brewer.pal(8, "Set2")
colLib <- colors[dds$group]
boxplot(log2(norm + 1), col = colLib, las = 2)

housekeeping <- c("Nono", "Eef2", "Rplp0", "Hprt", "Tbp", "Gapdh") #Housekeeping genes for distal colon, proximal colon.
controlgenes <- row.names(dds[rowData(dds)$gene_name %in% housekeeping, ])

ite <- min(8, nrow(colData(dds))-1) 
for(i in 1:ite) {
  n.k <- i
  s <- RUVg(round(norm), controlgenes, k=n.k)
  plotRLE(s$normalizedCounts, outline = FALSE, ylim = c(-2, 2), col=colLib, main = paste0("RUVg k=", n.k), las = 2)
  plotPCA(s$normalizedCounts, col=colLib, cex=1.2, main = paste0("RUVg k=", n.k))
}

n.k <- 4 #Distal colon all
n.k <- 4 #Distal colon female

s <- RUVg(round(norm), controlgenes, k=n.k)
plotRLE(s$normalizedCounts, outline = FALSE, ylim = c(-2, 2), col=colLib, main = paste0("RUVg k=", n.k), las = 2)
plotPCA(s$normalizedCounts, col=colLib, cex=1.2, main = paste0("RUVg k=", n.k))
s$W
all(rownames(s$W) == rownames(colData(dds)))
colData(dds) <- cbind(colData(dds), s$W)
d <- as.formula(paste0("~", paste(paste(paste0("W_", 1:n.k), collapse = " + "), "group", sep = " + ")))
d
design(dds) <- d

dds <- DESeq(dds)
res <- results(dds)
hist(res$pvalue)
summary(res)
DESeq2::plotMA(dds)

res$gene_name <- rowData(dds)$gene_name
res$geneID <- str_sub(rownames(res), 1, 18)
resOrd <- res[order(res$pvalue), ]
ihwRes <- ihw(pvalue ~ baseMean,  data = as.data.frame(resOrd), alpha = 0.1)
resOrd$ihw <- adj_pvalues(ihwRes)
table(resOrd$ihw <= 0.1)

result0.1 <- resOrd[resOrd$ihw <= 0.1, ]
plot(result0.1$log2FoldChange, -log10(result0.1$ihw), xlim = c(-4, 4), col = "red", pch = 16)
text(result0.1$log2FoldChange, -log10(result0.1$ihw), labels = result0.1$gene_name, cex = 0.8)

result15perfdr <- resOrd[resOrd$ihw <= 0.15, ] #For IBS vs Control
resup <- result15perfdr[result15perfdr$log2FoldChange > 0, ]
resdown <- result15perfdr[result15perfdr$log2FoldChange < 0, ]

group <- "Distal_colon"
gender <- "Females"
#group <- "DRGL4"
write.csv(result15perfdr, file = file.path("/data/users/sushmanagaraj/projects/rnaseq/IBS_GABA/analysis/Reanalysis_with_Salmon_quantification/IBS-vs-Control", group, gender, paste0(group, ".", gender, ".RUVg.DE_genes.padj15per.txt")))
write.csv(resOrd, file = file.path("/data/users/sushmanagaraj/projects/rnaseq/IBS_GABA/analysis/Reanalysis_with_Salmon_quantification/IBS-vs-Control", group, gender, paste0(group, ".", gender, ".RUVg.DE_analysis.txt")))
save(s, dds, file = file.path("/data/users/sushmanagaraj/projects/rnaseq/IBS_GABA/analysis/Reanalysis_with_Salmon_quantification/IBS-vs-Control", group, gender, "norm_matrix_dds.RData"))
```

#GO analysis
```{r,echo=FALSE,fig.height=10,fig.width=10}
fullresult <- result15perfdr
fullresult <- resup
fullresult <- resdown
library(DOSE)
library(clusterProfiler)
#library(pathview)
library(org.Mm.eg.db)
allOE_genes <- as.character(resOrd$geneID)
sigOE_genes <- as.character(fullresult$geneID)
ego <- enrichGO(gene = sigOE_genes, 
                universe = allOE_genes,
                keyType = "ENSEMBL",
                OrgDb = org.Mm.eg.db, 
                ont = "ALL", 
                pAdjustMethod = "BH", 
                qvalueCutoff = 0.05, 
                readable = TRUE)

cluster_summary <- data.frame(ego)
set <- ""
set <- "up"
set <- "down"
write.csv(cluster_summary, file.path("/data/users/sushmanagaraj/projects/rnaseq/IBS_GABA/analysis/Reanalysis_with_Salmon_quantification/IBS-vs-Control", group, gender, paste0("GO_analysis.", group, ".", gender, ".", set, ".csv")))
pdf(file.path("/data/users/sushmanagaraj/projects/rnaseq/IBS_GABA/analysis/Reanalysis_with_Salmon_quantification/IBS-vs-Control", group,  gender, paste0("GO_analysis_plots.", group, ".", gender, ".", set, ".pdf")), width = 15)
dotplot(ego, showCategory = 50)
x2 <- enrichplot::pairwise_termsim(ego) 
emapplot(x2)
#emapplot(ego, showCategory = 50)
#OE_foldchanges <- fullresult$log2FoldChange
#names(OE_foldchanges) <- fullresult$geneID
#cnetplot(ego, 
#         categorySize="pvalue", 
#         showCategory = 15, 
#         foldChange=OE_foldchanges, 
#         vertex.label.font=5)
dev.off()
```

#GSEA Analysis
```{r echo=FALSE, eval=FALSE}
#fgsea
library(biomaRt)
mart <- useMart(biomart = "ENSEMBL_MART_ENSEMBL", "dataset" = "mmusculus_gene_ensembl", host = "useast.ensembl.org")
bm <- getBM(attributes=c("ensembl_gene_id", "hsapiens_homolog_associated_gene_name"), filters = "ensembl_gene_id",  values = resOrd$geneID, mart=mart)
bm <- bm[!duplicated(bm$ensembl_gene_id), ]
bm$hsapiens_homolog_associated_gene_name[bm$hsapiens_homolog_associated_gene_name == ""] <- NA
resOrd$human <- bm[match(resOrd$geneID, bm$ensembl_gene_id), "hsapiens_homolog_associated_gene_name"]
res2 <- resOrd[,c("human", "stat")]
res2 <- res2[!is.na(res2$human), ]
res2 <- res2[!duplicated(res2$human), ]

library(fgsea)
ranks <- res2$stat
names(ranks) <- res2$human
pathways.hallmark <- gmtPathways("/data/users/sushmanagaraj/ref/MSigDB/h.all.v7.2.symbols.gmt")
fgseaRes <- fgsea(pathways=pathways.hallmark, stats=ranks)
fgseaRestidy <- fgseaRes[order(fgseaRes$NES, decreasing = TRUE), ]
pdf(file.path("/data/users/sushmanagaraj/projects/rnaseq/IBS_GABA/analysis/Reanalysis_with_Salmon_quantification/IBS-vs-Control/",  group, gender, paste0("fgsea_analysis_plots.hallmark_pathways.", group, ".", gender, ".pdf")))
ggplot(fgseaRestidy, aes(reorder(pathway, NES), NES)) + geom_col(aes(fill=padj<0.05)) + coord_flip() + labs(x="Pathway", y="Normalized Enrichment Score", title="Hallmark pathways NES from GSEA") + theme_minimal()
dev.off()


pathways.kegg <- gmtPathways("/data/users/sushmanagaraj/ref/MSigDB/c2.cp.kegg.v7.2.symbols.gmt")
keggRes <- fgsea(pathways = pathways.kegg, ranks)
keggRestidy <-  keggRes[order(keggRes$NES, decreasing = TRUE), ]
pdf(file.path("/data/users/sushmanagaraj/projects/rnaseq/IBS_GABA/analysis/Reanalysis_with_Salmon_quantification/IBS-vs-Control", group, gender, paste0("fgsea_analysis_plots.kegg_pathways.", group, ".", gender, ".pdf")), height = 20, width = 10)
ggplot(keggRestidy, aes(reorder(pathway, NES), NES)) + geom_col(aes(fill=padj<0.05)) + coord_flip() + labs(x="Pathway", y="Normalized Enrichment Score", title="Kegg pathways NES from GSEA") + theme_minimal()
dev.off()
#pdf("/data/users/sushmanagaraj/projects/rnaseq/IBS_GABA/analysis/troubleshooting/IBS-vs-Control/DRGT12/Females/fgsea_enrichment_plot.kegg_tgf_beta_signaling.DRGT12.females.normal.pdf")
#plotEnrichment(pathways.kegg[["KEGG_TGF_BETA_SIGNALING_PATHWAY"]], ranks) 
#plotEnrichment(pathways.kegg[["KEGG_AXON_GUIDANCE"]], ranks) + labs(title = "AXON_GUIDANCE")
#dev.off()
plotEnrichment(pathways.hallmark[["HALLMARK_TNFA_SIGNALING_VIA_NFKB"]], ranks) + labs(title = "HALLMARK_TNFA_SIGNALING_VIA_NFKB")

topPathwaysUp <- keggRes[ES > 0][head(order(pval), n=10), pathway]
topPathwaysDown <- keggRes[ES < 0][head(order(pval), n=10), pathway]
topPathways <- c(topPathwaysUp, rev(topPathwaysDown))
pdf(file.path("/data/users/sushmanagaraj/projects/rnaseq/IBS_GABA/analysis/Reanalysis_with_Salmon_quantification/IBS-vs-Control", group, gender, paste0("fgsea_gseatable.kegg_pathways.", group, ".", gender, ".pdf")), width = 15)
plotGseaTable(pathways.kegg[topPathways], ranks, keggRes, gseaParam=0.5)
dev.off()

topPathwaysUp <- fgseaRes[ES > 0][head(order(pval), n=10), pathway]
topPathwaysDown <- fgseaRes[ES < 0][head(order(pval), n=10), pathway]
topPathways <- c(topPathwaysUp, rev(topPathwaysDown))
pdf(file.path("/data/users/sushmanagaraj/projects/rnaseq/IBS_GABA/analysis/Reanalysis_with_Salmon_quantification/IBS-vs-Control", group, gender, paste0("fgsea_gseatable.hallmark_pathways.", group, ".", gender, ".pdf")), width = 15)
plotGseaTable(pathways.hallmark[topPathways], ranks, fgseaRes, gseaParam=0.5)
dev.off()

library(data.table)
fwrite(keggRestidy[keggRestidy$padj < 0.1, ], file=file.path("/data/users/sushmanagaraj/projects/rnaseq/IBS_GABA/analysis/Reanalysis_with_Salmon_quantification/IBS-vs-Control", group, gender, paste0("fgsea_analysis.kegg_pathways.", group, ".", gender, ".txt")), sep="\t", sep2=c("", " ", ""))
fwrite(fgseaRestidy[fgseaRestidy$padj < 0.1, ], file=file.path("/data/users/sushmanagaraj/projects/rnaseq/IBS_GABA/analysis/Reanalysis_with_Salmon_quantification/IBS-vs-Control", group, gender, paste0("fgsea_analysis.hallmark_pathways.", group, ".", gender, ".txt")), sep="\t", sep2=c("", " ", ""))
```

```{r}
resOrd[resOrd$gene_name %in% c("Cadm2", "Ncam1", "Bag6", "Phf2", "Fam120a", "Ckap2", "Dock9", "Tnxb", "Nos1", "Uchl1", "Snap25"), ]

```

#Plotting function
```{r}
load(here("Reanalysis_with_Salmon_quantification", "IBS-vs-Control", group, gender, "norm_matrix_dds.RData"))

#Prepare the normalized matrix
all(rownames(dds) == rownames(s$normalizedCounts))
table(duplicated(rowData(dds)$gene_name))
rowData(dds)$gene_name_uniq <- rowData(dds)$gene_name
rowData(dds)$gene_id_trimmed <- str_sub(rowData(dds)$gene_id, 1, 18)
rowData(dds)$gene_name_uniq[which(duplicated(rowData(dds)$gene_name))] <- paste0(rowData(dds)$gene_name[which(duplicated(rowData(dds)$gene_name))], "_",  rowData(dds)$gene_id_trimmed[which(duplicated(rowData(dds)$gene_name))])
rownames(s$normalizedCounts) <- rowData(dds)$gene_name_uniq

genes_to_plot <- c("Cadm2", "Ncam1", "Bag6", "Phf2", "Fam120a", "Ckap2", "Dock9") #IBS GWAS
pdf(here("Reanalysis_with_Salmon_quantification", "IBS-vs-Control", group, gender, "boxplots_of_IBS_GWAS_genes.pdf"))
plot_genes(genes_to_plot)
dev.off()

genes_to_plot <- c("Cdk18", "Snx24", "Mrpl22", "Lfng", "Calcb", "Bdnf", "Hmga2", "Arf2", "Fam227a") #Stool frequency GWAS
pdf(here("Reanalysis_with_Salmon_quantification", "IBS-vs-Control", group, gender, "boxplots_of_Stool_Freq_GWAS_genes.pdf"))
plot_genes(genes_to_plot)
dev.off()

plot_genes <- function(gene_to_plot) {
  #name1 <- sapply(gene_to_plot, function(x) { rownames(dds)[grep(x, rownames(dds))] })
  mydata <- as.data.frame(s$normalizedCounts[rownames(s$normalizedCounts) %in% gene_to_plot, ])
  mydata <- t(mydata)
  pmelt <- reshape2::melt(mydata)
  colnames(pmelt) <- c("sample", "gene", "normalized_count")
  pmelt$group <- colData(dds)[match(pmelt$sample, dds$Sample), "group"]
  ggplot(pmelt, aes(x = group, y = normalized_count)) + geom_boxplot(aes(fill = group), outlier.shape = NA) + geom_point(aes(x = group, y = normalized_count), position = "jitter", size = 1) + facet_wrap(pmelt$gene, scales = "free_y") + theme_classic() + theme(axis.text.x = element_text(angle = 90)) + ggtitle(paste0(group, " ", gender))
}
```

```{r echo=FALSE, eval=FALSE}
group <- "Distal_colon"
gender <- "Males"
path <- here("Reanalysis_with_Salmon_quantification", "IBS-vs-Control")
male <- read.csv(file.path(path, group, gender, paste0(group, ".", gender, ".RUVs.DE_genes.padj15per.txt")))
maleord <- male[order(male$pvalue), ]
#maletop50 <- maleord[1:50, "Gene_name"]

gender <- "Females"
female <- read.csv(file.path(path, group, gender, "RUVs", paste0(group, ".", gender, ".RUVs.DE_genes.padj15per.txt")))
femaleord <- female[order(female$pvalue), ]
#femaletop50 <- femaleord[1:50, "Gene_name"]

gender <- "combined"
all <- read.csv(file.path(path, group, gender, "RUVs", paste0(group, ".", gender, ".RUVs.DE_genes.padj15per.txt")))
allord <- all[order(all$pvalue), ]
#alltop50 <- allord[1:50, "Gene_name"]

#library(VennDiagram)
#Proximal and Distal Colon.
#venn.diagram(list(male = maletop50, female = femaletop50, all = alltop50), main = "Ileum vs Duodenum Top 50 genes", filename = #"/data/users/sushmanagaraj/projects/rnaseq/IBS_GABA/analysis/troubleshooting/venn.ileum-vs-duodenum.top50.tiff")
#venn.diagram(list(male = maleord$Gene_name, female = femaleord$Gene_name, all = allord$Gene_name), main = "Distal Colon vs. Ileum all DE #genes (FDR <= 0.1, LFC > |0.5|)", filename = #"/data/users/sushmanagaraj/projects/rnaseq/IBS_GABA/analysis/troubleshooting/venn.distal_colon-vs-ileum.all_DE_genes.tiff")

#mtop50 <- data.frame(gene_name = maletop50, analysis = rep("males_only", length(maletop50)))
#ftop50 <- data.frame(gene_name = femaletop50, analysis = rep("females_only", length(femaletop50)))
#atop50 <- data.frame(gene_name = alltop50, analysis = rep("combined", length(alltop50)))

colnames(maleord) <- paste0("males.", colnames(maleord))
colnames(maleord)[1] <- "GeneID"
colnames(femaleord) <- paste0("females.", colnames(femaleord))
colnames(femaleord)[1] <- "GeneID"
colnames(allord) <- paste0("combined.", colnames(allord))
colnames(allord)[1] <- "GeneID"

merged <- Reduce(function(x, y) merge(x, y, all = TRUE, by = "GeneID"), list(maleord, femaleord, allord))
write.csv(merged, file = file.path(path, group, "Distal_colon.all_analyses_combined.RUVs.txt"))

```

#Annotation
```{r}
merged <- read.csv(file.path(path, group, "Distal_colon.all_analyses_combined.RUVs.txt"))
merged$gene_name <- rowData(dds.full)[match(merged$GeneID, rowData(dds.full)$gene_id), "gene_name"]

#IBS biomarkers
biomarkers <- read.csv(here("data", "IBS_biomarkers.txt"), sep = ",", header = TRUE)
merged$IBS_biomarker <- biomarkers[match(merged$gene_name, biomarkers$Gene), "Type"]

#Neuronal genes
source("/data/users/sushmanagaraj/ref/genelists/Brown_Goff_hand-curated_gene_sets_2018.R")

sets <- list(All_channels, Amine_receptors, Axon_genes, CaChannels, ClChannels, Gage_FOS_dn, Gage_FOS_up, GluR_genes, GPCRs, Gray.activity.genes, Ion_channels, K_channels_Mouse, KChannel_reg_subunits, KChannels, Ligand_gated_ion_channels, Ligand_gated_ion_channels_HUGO, LTP_genes, Na_channels_Mouse, NaChannels, Neuropeptides, Other_receptors, Peptide_receptors, Synaptic_transmission_genes, VIP_calcitonin, Voltage_gated_channels_HUGO, gap_junctions, purine_signalling)
names(sets) <- c("All_channels", "Amine_receptors", "Axon_genes", "CaChannels", "ClChannels", "Gage_FOS_dn", "Gage_FOS_up", "GluR_genes", "GPCRs", "Gray.activity.genes", "Ion_channels", "K_channels_Mouse", "KChannel_reg_subunits", "KChannels", "Ligand_gated_ion_channels", "Ligand_gated_ion_channels_HUGO", "LTP_genes", "Na_channels_Mouse", "NaChannels", "Neuropeptides", "Other_receptors", "Peptide_receptors", "Synaptic_transmission_genes", "VIP_calcitonin", "Voltage_gated_channels_HUGO", "gap_junctions", "purine_signalling")

neuronal.temp1 <- unlist(sets)
neuronal.df <- as.data.frame(neuronal.temp1)
neuronal.df$Annotation <- rownames(neuronal.df)
neuronal.df$Annotation <- gsub("[[:digit:]]+", "", neuronal.df$Annotation)
neuronal.df2 <- aggregate(data=neuronal.df,Annotation~.,FUN=paste,collapse=",")

merged$Neuronal_gene <- neuronal.df2[match(merged$gene_name, neuronal.df2$neuronal.temp1), "Annotation"]

library(biomaRt)
ensembl = useEnsembl(biomart="ensembl", dataset="mmusculus_gene_ensembl")
genedesc <- getBM(attributes=c('external_gene_name','description'), filters = 'external_gene_name', values = merged$gene_name, mart =ensembl)
merged$description <- genedesc[match(merged$gene_name, genedesc$external_gene_name), "description"]

write.table(merged, file = file.path(path, group, "Distal_colon.all_analyses_combined.annotations.txt"), sep = "\t")
```